// @n8n/api-types/dist/chat-hub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chatHubMessagesResponseSchema = exports.chatHubConversationsResponseSchema = exports.chatHubConversationSchema = exports.chatHubMessageSchema = exports.ChatHubSendMessageRequest = exports.chatModelsRequestSchema = exports.chatHubConversationModelSchema = exports.PROVIDER_CREDENTIAL_TYPE_MAP = exports.chatHubProviderSchema = void 0;
const zod_1 = require("zod");
const zod_class_1 = require("zod-class");
exports.chatHubProviderSchema = zod_1.z.enum(['openai', 'anthropic', 'google']);
exports.PROVIDER_CREDENTIAL_TYPE_MAP = {
    openai: 'openAiApi',
    anthropic: 'anthropicApi',
    google: 'googlePalmApi',
};
exports.chatHubConversationModelSchema = zod_1.z.object({
    provider: exports.chatHubProviderSchema,
    model: zod_1.z.string(),
});
exports.chatModelsRequestSchema = zod_1.z.object({
    credentials: zod_1.z.record(exports.chatHubProviderSchema, zod_1.z.string().nullable()),
});
class ChatHubSendMessageRequest extends zod_class_1.Z.class({
    messageId: zod_1.z.string().uuid(),
    sessionId: zod_1.z.string().uuid(),
    message: zod_1.z.string(),
    model: exports.chatHubConversationModelSchema,
    credentials: zod_1.z.record(zod_1.z.object({
        id: zod_1.z.string(),
        name: zod_1.z.string(),
    })),
}) {
}
exports.ChatHubSendMessageRequest = ChatHubSendMessageRequest;
exports.chatHubMessageSchema = zod_1.z.object({
    id: zod_1.z.string().uuid(),
    conversationId: zod_1.z.string().uuid(),
    role: zod_1.z.enum(['user', 'assistant']),
    content: zod_1.z.string(),
    createdAt: zod_1.z.string().datetime(),
});
exports.chatHubConversationSchema = zod_1.z.object({
    id: zod_1.z.string().uuid(),
    title: zod_1.z.string(),
    createdAt: zod_1.z.string().datetime(),
    updatedAt: zod_1.z.string().datetime(),
});
exports.chatHubConversationsResponseSchema = zod_1.z.array(exports.chatHubConversationSchema);
exports.chatHubMessagesResponseSchema = zod_1.z.array(exports.chatHubMessageSchema);
